slow_query_exporter.rb

```diff
@@ -0,0 +1,266 @@
+    end
+    gelf_attrs
+  end
+
+  private
+
+  # This is just a straight port of pt-query-digest's Perl fingerprint function.
+  # The ugly regexps aren't my fault! :-D
+  def normalized_query
+    q = attributes[:query_string].dup
+    return "mysqldump" if q =~ %r{\ASELECT /\*!40001 SQL_NO_CACHE \*/ \* FROM `}
+    return "percona-toolkit" if q =~ %r{/\*\w+\.\w+:[0-9]/[0-9]\*/}
+    return q if q =~ /\Aadministrator command: /i
+    return $1.downcase if q =~ /\A\s*(call\s+\S+)\(/i
+
+    # Truncate multi-value INSERT statements
+    q = $1 if q =~ /\A((?:INSERT|REPLACE)(?: IGNORE)?\s+INTO.+?VALUES\s*\(.*?\))\s*,\s*\(/im
+
+    q.gsub!(%r{/\*[^!].*?\*/}m, "")                     # Strip multi-line comments
+    q.gsub!(%r{(?:--|#)[^'"\r\n]*(?=[\r\n]|\Z)}m, "")   # Strip single-line comments
+    return "use ?" if q =~ /\Ause \S+\Z/i
+
+    # Replace quoted strings with "?"
+    q.gsub!(/\\["']/, "")
+    q.gsub!(/".*?"/, "?")
+    q.gsub!(/'.*?'/, "?")
+
+    q.downcase!
+    q.gsub!(/\bfalse\b|\btrue\b/, "?")      # Replace boolean values with "?"
+    q.gsub!(/[0-9+-][0-9a-f.xb+-]*/, "?")   # Replace numbers with "?"
+    q.gsub!(/[xb.+-]\?/, "?")               # Strip number prefixes
+
+    q.gsub!(/\bnull\b/, "?")   # Replace NULLs with "?"
+
+    q.gsub!(/\b(in|values?)(?:[\s,]*\([\s?,]*\))+/, "\\1(?+)")                   # Collapse IN and VALUES lists
+    q.gsub!(/\b(select\s.*?)(?:(\sunion(?:\sall)?)\s\1)+/, "\\1 /*repeat$2*/")   # Collapse UNIONs
+    q.sub!(/\blimit \?(?:, ?\?| offset \?)?/, "limit ?")   # LIMITs and OFFSETs
+    q.gsub(/\border by (.+?)\s+asc/, "order by \\1")       # Remove extraneous ASCs from ORDER BYs
+  end
+
+  # Remove leading/trailing whitespace and compress all other runs of whitespace into a single space.
+  def compress(str)
+    str.strip.tr_s(" \n\t\r\f", " ")
+  end
+end
+
+
+$graylog_host = "localhost"
+$graylog_port = 12201
+$delay = 0.1
+$verbose = false
+$foreground = false
+
+HELP_TEXT = "Usage: slow_query_exporter.rb [-fv] [-d delay] [-h host] [-p port] slow_query.log
+Options:
+    -h, --host         The Graylog host
+    -p, --port         The Graylog port
+    -d, --delay        An interval to pause after each GELF message, in (possibly fractional) seconds
+    -v, --verbose      Print entries to stdout as they're parsed
+    -f, --foreground   Don't daemonize on startup
+    -?, --help         Display this help text
+"
+
+OptionParser.new do |opts|
+  opts.banner = HELP_TEXT
+
+  opts.on("-h", "--host") { |host| $graylog_host = host }
+  opts.on("-p", "--port") { |port| $graylog_port = port }
+  opts.on("-d", "--delay") { |delay| $delay = delay.to_f }
+  opts.on("-v", "--verbose") { $verbose = true }
+  opts.on("-f", "--foreground") { $foreground = true }
+  opts.on("-?", "--help") do
+    $stderr.puts(HELP_TEXT)
+    exit
+  end
+end.parse!
+
+if ARGV.empty?
+  $stderr.puts(HELP_TEXT)
+  exit 1
+end
+
+Daemons.daemonize(dir_mode: :normal, dir: RUN_DIRECTORY) unless $foreground
```

Comment:
`Daemons.daemonize` doesn't do the stop/restart/start behaviour, it just daemonizes. But if I can get the same behaviour in Ruby core now, I'll just do that. Good idea.